File: bookstore-service-micronaut/src/main/java/example/micronaut/bookstore/core/service/BookService.java
package example.micronaut.bookstore.core.service;

import example.micronaut.bookstore.api.kafka.producer.BookProducer;
import example.micronaut.bookstore.api.model.Book;
import example.micronaut.bookstore.api.model.BookRequest;
import example.micronaut.bookstore.api.model.search.BookSearchCriteria;
import example.micronaut.bookstore.api.model.search.PagedSearchResult;
import example.micronaut.bookstore.core.mapper.EntityMapper;
import example.micronaut.bookstore.persistence.entity.BookEntity;
import example.micronaut.bookstore.persistence.repository.BookRepository;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;

import javax.transaction.Transactional;
import java.util.UUID;

@Singleton
public class BookService {

    @Inject
    private BookRepository bookRepository;

    @Inject
    private EntityMapper entityMapper;

    @Inject
    private BookProducer bookProducer;

    @Transactional
    public PagedSearchResult<Book> getBooks(BookSearchCriteria criteria) {
        return PagedSearchResult.from(bookRepository.findAll(criteria, criteria.toPageable()), entityMapper::map);
    }

    public Book getBook(Long bookId) {
        return entityMapper.map(bookRepository.findById(bookId)
                .orElseThrow(() -> new IllegalArgumentException("Book with ID %s does not exist".formatted(bookId))));
    }

    public Book createBook(BookRequest request) {
        BookEntity book = entityMapper.map(request);
        Book newBook = entityMapper.map(bookRepository.save(book));
        bookProducer.sendBook(UUID.randomUUID().toString(), newBook);
        return newBook;
    }

    public Book updateBook(Long bookId, BookRequest request) {
        BookEntity book = bookRepository.findById(bookId)
                .orElseThrow(() -> new IllegalArgumentException("Book with ID %s does not exist".formatted(bookId)));
        entityMapper.update(book, request);

        return entityMapper.map(bookRepository.update(book));
    }
}

File: bookstore-service-micronaut-kotlin/src/main/kotlin/com/example/micronaut/bookstore/service/BookService.kt
package com.example.micronaut.bookstore.service

import com.example.micronaut.bookstore.controller.model.Book
import com.example.micronaut.bookstore.controller.model.CreateBookRequest
import com.example.micronaut.bookstore.kafka.producer.BookProducer
import com.example.micronaut.bookstore.kafka.producer.event.KafkaEvent
import com.example.micronaut.bookstore.repository.BookRepository
import com.example.micronaut.bookstore.service.model.BookSearchCriteria
import io.micronaut.data.model.Page
import jakarta.inject.Singleton
import jakarta.transaction.Transactional

@Singleton
open class BookService(private val bookRepository: BookRepository, private val bookProducer: BookProducer) {

    @Transactional
    open fun getBooks(bookSearchCriteria: BookSearchCriteria): Page<Book> {
        return bookRepository.findAll(bookSearchCriteria, bookSearchCriteria.toPageable())
            .map { Book.from(it) }
    }

    fun createBook(createBookRequest: CreateBookRequest): Book {
        val book = Book.from(bookRepository.save(createBookRequest.toEntity()))
        publishBookEvent(book)
        return book
    }

    private fun publishBookEvent(book: Book) {
        bookProducer.sendBook(book.id, KafkaEvent.of(book))
    }
}